packages/cli/src/active-executions.ts

```diff
@@ -193,32 +205,33 @@ export class ActiveExecutions {
 	}
 
 	setStatus(executionId: string, status: ExecutionStatus) {
-		this.getExecution(executionId).status = status;
+		this.getExecutionOrFail(executionId).status = status;
 	}
 
 	getStatus(executionId: string): ExecutionStatus {
-		return this.getExecution(executionId).status;
+		return this.getExecutionOrFail(executionId).status;
 	}
 
 	/** Wait for all active executions to finish */
 	async shutdown(cancelAll = false) {
-		let executionIds = Object.keys(this.activeExecutions);
+		const { withResponsePromise, others } = this.groupExecutionIds();
+		const executionIdsToCancel = [...withResponsePromise];
 
 		if (config.getEnv('executions.mode') === 'regular') {
 			// removal of active executions will no longer release capacity back,
 			// so that throttled executions cannot resume during shutdown
 			this.concurrencyControl.disable();
-		}
 
-		if (cancelAll) {
-			if (config.getEnv('executions.mode') === 'regular') {
-				await this.concurrencyControl.removeAll(this.activeExecutions);
+			if (cancelAll) {
+				executionIdsToCancel.push(...others);
 			}
-
-			executionIds.forEach((executionId) => this.stopExecution(executionId));
 		}
 
+		executionIdsToCancel.forEach((executionId) => this.stopExecution(executionId));
+		await this.concurrencyControl.removeAll(executionIdsToCancel);
+
 		let count = 0;
+		let executionIds = Object.keys(this.activeExecutions);
```

Comment:
What if new executions end up in the active executions between when we cancel them and we start waiting for them? I guess those never get cancelled?

Reply:
we already have some mechanisms to prevent that from happening, but we still might need more checks. however, that's not a problem specific to this PR, right?

Reply:
No. I'm just pointing out potential issues we have in the current implementation

Reply:
should we add a `isShutingDown` flag here, like we have on a couple of other services?